feat: add support for langfuse data format

whanou opened the pull request:
This is actually an MVP version, i'm figuring out a way to get (is it possible to export it maybe?) input/output and attributes. I see them on Langfuse' website, but they are missing in exported JSON. I'm thinking whether the only way for it to be seen is a fetch. Most likely not.

Also instead of writing an adapter from langfuse to otlp, i decided to implement adapter to our types straight away. There's not reason imo for this intermediate step since it adds a bunch of conversion logic which we don't need at all. @mikhin lmk what you think of this 

<img width="843" height="485" alt="image" src="https://github.com/user-attachments/assets/c806b2c8-c3f9-478e-a15d-be28c37e38fb" />

mikhin commented on packages/data/src/services/convert-langfuse-observation-to-span-card.ts:
```diff
@@ -0,0 +1,80 @@
+  return observation.costDetails?.total || 0;
+}
+
+function getInputOutput(): InputOutputData {
```
I think the time has come.

Could I ask you to refactor this PR and split the adapters into directories? So that the lib would export one main function for each adapter. At the same time, inside, the common utility functions could be shared between adapters, and the unique functions would be in each adapter directory.

If you have any questions, let's discuss it.

whanou replied:
Agree, makes sense. Especially since new providers (such as langchain) will come in foreseeable future

whanou commented:
Pushed some work for visibility, but there are a bunch of stuff yet to do. I though of the best approach an decided that the best idea is to make big overhaul of the logic and architecture. Here are my ideas:

- We need some basic entity which describes what adapter does and which can be extended to create different adapters. Preferably without creating a lot of functions/utils which bear similar names
- We need more strict separation for OpenTelemetry and Langfuse types and logic

Here's what i come up with:
- There are 3 folders inside `services` now - common, open-telemetry and langfuse
- common folder has some common utils and `adapter.ts` which has abstract class with some basic interfaces. Each of otl and langfuse (and potential new adapters in the future) will extend this class
- Why the class? First - it's really good for connecting logic and showing hierarchy. Second - we expect folks to have experience with python so having classes if kinda familiar
- Each service also has a folder named `tests` which test a part of the functionality of the class. This is easier to grasp imo

So the structure is like this:
```
services
  common
    tests
      util-1.test.ts
      util-2.test.ts
    util-1.ts
    util-2.ts
    adapter.ts
  open-telemetry
    tests
      feature-1.test.ts
      feature-2.test.ts
    adapter.ts
  langfuse
     tests
      feature-1.test.ts
      feature-2.test.ts
    adapter.ts
```

What i did to achieve this:
- Rewrote everything to be a class
- Moved the tests and rewrote some of them
- Renamed otlp types and updated them everywhere. I wanted to achieve stronger separation - before we had a `Span` type for example - but this same type can not be used with Langfuse, so i renamed it to `OpenTelemetrySpan`
- Updated usage in stories, examples etc
- Added "Span" status (it is used in Langfuse)

What is left to do:
- Take langfuse logic and do the same
- Update readme and docs
- Check all the places to make sure types and logic between adapters are separated correctly
- Add missing tests

@mikhin WDYT about this approach? Imo it allows for easier extensibility. Not only that - it also makes it super duper obvious how to add new adapters

mikhin commented on packages/data/src/services/open-telemetry/adapter.ts:
```diff
@@ -0,0 +1,443 @@
+
+import { SpanAdapter } from "../common/adapter";
+
+export class OpenTelemetrySpanAdapter extends SpanAdapter<
```
The naming and semantics here are a bit confusing right now. `SpanAdapter` doesn’t clearly convey its purpose, and the generics aren’t self-explanatory. What do they represent? Should they be input/output types? If so, shouldn’t the output be `UiSpan` instead of `OpenTelemetrySpan`? It would be great if the class and its generics were more descriptive so the intent is clear at a glance.

whanou commented on packages/data/src/services/open-telemetry/adapter.ts:
```diff
@@ -0,0 +1,443 @@
+  };
+}
+
+export const openTelemetrySpanAdapter = new OpenTelemetrySpanAdapter();
```
Yep, i agree - we underutilize oop possibilities (and we don't need any of them). I'll think of a better way to structure it so that it offers separation of concerns with clear hierarchy, while being smaller and simpler

whanou commented on packages/types/src/types/index.ts:
```diff
@@ -20,7 +20,7 @@ export type TraceSpan = {
   type: TraceSpanCategory;
   raw: string;
-  attributes: SpanAttribute[];
+  attributes?: OpenTelemetrySpanAttribute[];
```
I don't think TraceRecord should be generic, but rather we should move attributes to base types. I did that - now we need to (if required) map whatever is in the raw data to our attributes

whanou commented:
@mikhin Thanks for the review! I made some adjustments:

- Now adapters are just plain objects which implement `SpanAdapter` interface
- There is a clear separation - objects methods are something we _want_ to get exposed. And utils (placed in designated folder) are some local stuff (separation of code, small utils, internals etc.). So i made some of the functions methods because _i think_ this is what might be needed in folks code - not only parsing documents are returning `TraceSpan[]`, but also some smaller stuff like getting cost, or mapping single span (we assume a single span can be fetched and loaded into the app)
- Also - methods for mapping input documents/spans are called `convertRawXToY` - to emphasize we are dealing with something unprocessed
- Also i updated types for Langfuse, now they are more accurate

There are few things which i think we need to address later:
- I haven't found good-fitting badges for some of the Langfuse observation types (such as guardrail, span and event), so i extended our types. I think it's fine for now, but with new adapters being added - we will probably pollute our types more and more
- It seems like Langfuse has no `status` field (it's always successful?). I will ask Gleb whether this needs to be displayed differently. For now i always return "success"
- It seems like you need to make a request to Langfuse API to get full observation data (i'm only interested in input/output). Exported JSON has no input/output 🤔 I'm figuring a way to get this data - in real apps this does not seem like a problem - everyone will be fetching data anyway. But for our demos - we would probably want this data to be present. If i won't find better way - i'll just fill it by hands with copy-pasting

mikhin commented:
@whanou, it looks awesome, thank you very much!
